### PR TITLE
HV-128: Add New Observables

### PR DESCRIPTION
**Pull request by timmccall-wf**
Problem
-------

There is no way for consumers to reliably detect when scale or scroll position/translation will change for both the list and item maps.

Solution
--------

Add the following observables:

AwesomeMap:
- onScaleWillChange
- onTranslationWillChange

ScrollList:
- onScaleWillChange
- onScrollPositionWillChange
- onItemScrollPositionChanged
- onItemScrollPositionWillChange

Unit Tests
----------

Added

How To +10/QA
-------------

- Instructions in upstream changes.

@todbachman-wf

### CODE REVIEW

**Comment by rmconsole-wf** [comment]
When this pull is merged I will use the following information:
		Version: wf-uicomponents 1.0.2
		Release Ticket: HY-1818

***
*Last updated on Thursday, January 22 09:27 PM CST*

**Comment by patkujawa-wf** [comment]
Why does only one observable have a 'Changed' version? I would imagine that consumers would be just as interested in an observable when something finished if they are interested when it starts to change, right?

**Comment by patkujawa-wf** [comment]
React uses the Will/Did terminology, e.g. WillChange/DidChange. .NET uses the Changing/Changed terminology. Any reason to diverge with this mix of WillChange/Changed?

**Comment by timmccall-wf** [comment]
Part is to not refactor the existing, the other part is that I don't like gerunds because the things aren't changing at present, but they will change. But I can change to the gerund if it makes more sense.

File: src/awesome_map/AwesomeMap.js [hunk 1]
```diff
@@ -148,6 +148,18 @@ define(function(require) {
...
+         * Observable for subscribing to pending scale changes.
+         *
+         * @method AwesomeMap#onScaleWillChange
+         * @param {Function} callback
+         *        Invoked with (sender, {
+         *            currentScale: {number}
+         *            nextScale: {number}
+         *        })
+         */
+        this.onScaleWillChange = Observable.newObservable();
```
Comment (patkujawa-wf) [inline src/awesome_map/AwesomeMap.js]: In my experience, most observables/events having to do with something that is going to happen usually allow the consumers to override what is going to happen. So can consumers change `nextScale`? Would we want to let them?

File: src/awesome_map/TransformationQueue.js [hunk 2]
```diff
@@ -200,6 +203,35 @@ define(function(require) {
...
+                map.onScaleWillChange.dispatch([map, {
+                    event: event,
+                    currentScale: currentState.scale,
+                    nextScale: nextState.scale
+                }]);
+            }
+            if (currentState.translateX !== nextState.translateX ||
+                currentState.translateY !== nextState.translateY) {
+                map.onTranslationWillChange.dispatch([map, {
+                    event: event,
```
Comment (patkujawa-wf) [inline src/awesome_map/TransformationQueue.js]: Do you want to document the event prop? I didn't see that in the docstring.

File: src/scroll_list/ScrollList.js [hunk 3]
```diff
@@ -178,28 +178,6 @@ define(function(require) {
...
-         * Observable for subscribing to scroll to events.
-         * This event occurs when the resulting call to scrollToItem
-         * finishes its animation.
-         *
-         * @method ScrollList#onScrollToItemStarted
-         * @param {Function} callback
-         *        Invoked with no parameters
-         */
-        this.onScrollToItemStarted = Observable.newObservable();
-
```
Comment (patkujawa-wf) [inline src/scroll_list/ScrollList.js]: Who was consuming these? Do we need a major version bump now?

**Comment by patkujawa-wf** [comment]
Oh, I see, onScaleChanged already existed, so this is just adding the willChange for parity.

I think it'll be less cognitive load to stick with a familiar convention for changing/changed/willChange. Since we're in JS and not C#, I'd go with React's convention, personally.

**Comment by patkujawa-wf** [comment]
+1 unless you want to make the suggested changes @timmccall-wf

File: src/awesome_map/AwesomeMap.js (hunk 1 above)
Comment (timmccall-wf) [inline src/awesome_map/AwesomeMap.js]: Interceptors rely on being able to do this during `onTransformStarted`, so yeah, I think it's fine.

File: src/awesome_map/TransformationQueue.js (hunk 2 above)
Comment (timmccall-wf) [inline src/awesome_map/TransformationQueue.js]: Removing...

**Comment by timmccall-wf** [comment]
@patkujawa-wf Changed.

**Comment by patkujawa-wf** [comment]
+1 thanks!

**Comment by patkujawa-wf** [comment]
+1 nice changes to the scroll list

**Comment by tomconnell-wf** [comment]
+10 - I tested a whole bunch of stuff while zoomed in.  Annotation interactions, clicks, drags, etc etc.

**Comment by charliekump-wf** [comment]
@rmconsole-wf

**Comment by charliekump-wf** [comment]
QA Resource Approval: +1

- [x] Testing instruction
- [x] Dev +1's
- [x] Dev/QA +10 with detail of what was tested
- [x] Unit test created/updated  
- [x] All unit tests pass 

Merging into master.

**Pull request merged**